control/controlclient/noise.go

```diff
@@ -295,11 +314,35 @@ func (nc *NoiseClient) getConn(ctx context.Context) (*noiseConn, error) {
 	}
 	nc.mu.Unlock()
 
-	conn, err, _ := nc.sfDial.Do(struct{}{}, nc.dial)
-	if err != nil {
-		return nil, err
+	for {
+		// We singeflight the dial to avoid making multiple connections, however
+		// that means that we can't simply cancel the dial if the context is
+		// canceled. Instead, we have to additionally check that the context
+		// which was canceled is our context and retry if our context is still
+		// valid.
+		conn, err, _ := nc.sfDial.Do(struct{}{}, func() (*noiseConn, error) {
+			c, err := nc.dial(ctx)
+			if err != nil {
+				if ctx.Err() != nil {
+					return nil, contextErr{err}
```

Comment:
there's a race here: maybe the dial failed due to getting a RST from the peer. But then independently, by chance, the context was canceled just before line 326's check. Now you're wrapping a "connection refused" error in a "contextErr{...}" wrapper.

That's harmless for what you need, but doesn't match what the docs on L290 say.

Reply:
Just do `contextErr{ctx.Err()}` instead?

Reply:
yeah, i wasn't sure how to handle the race. `contextErr{ctx.Err()}` makes sense. Done.